README.md
# React gradient color picker

This is being published publicly to make my life a little eaiser using the previous version than
linking my fixed/updated via github.

**USE AT YOUR OWN RISK. SUBJECT TO CHANGES AT MY WHIM.**

### Inspired by [gpickr](https://github.com/Simonwep/gpickr)

### Inspired by [react-gcolor-picker](https://github.com/undind/react-gcolor-picker)

[![NPM](https://img.shields.io/npm/v/react-gcolor-picker-2.svg)](https://www.npmjs.com/package/react-gcolor-picker-2) [![License](https://img.shields.io/npm/l/react-gcolor-picker-2.svg)](https://github.com/margolisj/react-gcolor-picker/blob/main/LICENSE)

**[Demo](https://margolisj.github.io/react-gcolor-picker/)**

## Install

### **Important: this component uses React Hooks and works on React version 16.8.0 and higher**

---

```bash
npm install --save react-gcolor-picker-2
```

OR

```bash
yarn add react-gcolor-picker-2
```

## Usage

```tsx
import React from 'react';

import ReactGPicker from 'react-gcolor-picker-2';

function App() {
  const onChange = (value) => {
    console.log(value);
  };

  return <ReactGPicker value='red' onChange={onChange} />;
}

export default App;
```

## Props

| Attribute            |    Type    |           Default           | Description                                                                                           |
| :------------------- | :--------: | :-------------------------: | :---------------------------------------------------------------------------------------------------- |
| value                |  `string`  |          `#ffffff`          | Default color value. Accept: rgba/rgb, hsla/hsl, named colors                                         |
| format               |  `string`  |            `rgb`            | Return value format. Accept: 'rgb', 'hex', 'hsl'                                                      |
| gradient             |   `bool`   |           `false`           | Show gradient color panel                                                                             |
| solid                |   `bool`   |           `true`            | Show solid color panel                                                                                |
| debounceMS           |  `number`  |            `300`            | Debounce ms value                                                                                     |
| debounce             |   `bool`   |           `true`            | Debouce off/on                                                                                        |
| showAlpha            |   `bool`   |           `true`            | Show/hide alpha input and range                                                                       |
| showInputs           |   `bool`   |           `true`            | Show/hide inputs alpha and color                                                                      |
| showGradientResult   |   `bool`   |           `true`            | Show/hide gradient result fields(angle, mode, position)                                               |
| showGradientStops    |   `bool`   |           `true`            | Show/hide gradient color stops                                                                        |
| showGradientMode     |   `bool`   |           `true`            | Show/hide gradient color mode switcher                                                                |
| showGradientAngle    |   `bool`   |           `true`            | Show/hide gradient angle for linear-gradient                                                          |
| showGradientPosition |   `bool`   |           `true`            | Show/hide gradient position for radial-gradient                                                       |
| popupWidth           |  `number`  |            `267`            | Popup width                                                                                           |
| colorBoardHeight     |  `number`  |            `120`            | Board color height                                                                                    |
| defaultColors        |  `array`   | [List](#default-color-list) | Default colors array for panel picker                                                                 |
| defaultActiveTab     |  `string`  |         `undefined`         | Default value for active tab when initializing the component, takes two values: `solid` or `gradient` |
| onChangeTabs         | `function` |           `null`            | Default onChange function detect when tabs change and return one of the values: `solid` or `gradient` |
| onChange             | `function` |           `null`            | Default onChange function returns string value in the given format                                    |

When passing a value for a gradient, you must specify the position of all colors. Otherwise the component will throw an exception.
For example:

### Wrong

```
linear-gradient(180deg, #000000,#ff0000)
```

### Correct

```
linear-gradient(180deg, #000000 0%,#ff0000 100%)
```

If you are using a radial gradient a list of possible directions for it:

```
circle at left top
circle at center top
circle at right top
circle at left
circle at center
circle at right
circle at left bottom
circle at center bottom
circle at right bottom
```

## Default color list

```
'#FF6900',
'#FCB900',
'#7BDCB5',
'#00D084',
'#8ED1FC',
'#0693E3',
'#ABB8C3',
'#607d8b',
'#EB144C',
'#F78DA7',
'#ba68c8',
'#9900EF',
'linear-gradient(0deg, rgb(255, 177, 153) 0%, rgb(255, 8, 68) 100%)',
'linear-gradient(270deg, rgb(251, 171, 126) 8.00%, rgb(247, 206, 104) 92.00%)',
'linear-gradient(315deg, rgb(150, 230, 161) 8.00%, rgb(212, 252, 121) 92.00%)',
'linear-gradient(to left, rgb(249, 240, 71) 0%, rgb(15, 216, 80) 100%)',
'linear-gradient(315deg, rgb(194, 233, 251) 8.00%, rgb(161, 196, 253) 92.00%)',
'linear-gradient(0deg, rgb(0, 198, 251) 0%, rgb(0, 91, 234) 100%)',
'linear-gradient(0deg, rgb(167, 166, 203) 0%, rgb(137, 137, 186) 51.00%, rgb(137, 137, 186) 100%)',
'linear-gradient(0deg, rgb(80, 82, 133) 0%, rgb(88, 94, 146) 15.0%, rgb(101, 104, 159) 28.00%, rgb(116, 116, 176) 43.00%, rgb(126, 126, 187) 57.00%, rgb(131, 137, 199) 71.00%, rgb(151, 149, 212) 82.00%, rgb(162, 161, 220) 92.00%, rgb(181, 174, 228) 100%)',
'linear-gradient(270deg, rgb(255, 126, 179) 0%, rgb(255, 117, 140) 100%)',
'linear-gradient(90deg, rgb(120, 115, 245) 0%, rgb(236, 119, 171) 100%)',
'linear-gradient(45deg, #2e266f 0.00%, #9664dd38 100.00%)',
'radial-gradient(circle at center, yellow 0%, #009966 50%, purple 100%)'
```

## FAQ

**Q:** There's any possibility to remove extra gradient lines?

**A:** Yes, you only need to drag them outside(bottom) the Gradient component area or double click on the element you want to remove.

## License

MIT ©


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [0.0.9](https://github.com/margolisj/react-gcolor-picker/compare/0.0.8...0.0.9)

- More rollup fixes [`f1c3756`](https://github.com/margolisj/react-gcolor-picker/commit/f1c3756d8183adf4ea0c000387bed4ab3f19c4b8)
- Scripts. [`1649a3e`](https://github.com/margolisj/react-gcolor-picker/commit/1649a3eff010489043e4689d0d819556b54dc77b)
- Lint fix. [`b73a590`](https://github.com/margolisj/react-gcolor-picker/commit/b73a590dfc5741aeeb1e4249468c9707a39592ad)

#### [0.0.8](https://github.com/margolisj/react-gcolor-picker/compare/0.0.7...0.0.8)

> 2 August 2023

- Simple one to try. [`899ee7b`](https://github.com/margolisj/react-gcolor-picker/commit/899ee7bf40e4082c1943f1b7188f4656b5198cc4)
- Lint fix. [`0a08de7`](https://github.com/margolisj/react-gcolor-picker/commit/0a08de76dba4e0d9347c54f6d03064731a948772)
- Release v0.0.8 [`742e11c`](https://github.com/margolisj/react-gcolor-picker/commit/742e11ce2c2a69f1d17329d4c9cfb0825823c61b)

#### [0.0.7](https://github.com/margolisj/react-gcolor-picker/compare/0.0.6...0.0.7)

> 2 August 2023

- Roll Up TS Bug [`#8`](https://github.com/margolisj/react-gcolor-picker/pull/8)
- Release v0.0.7 [`592f6d5`](https://github.com/margolisj/react-gcolor-picker/commit/592f6d5e0c0aeebe5ccfbeacce7669b3990debb9)

#### [0.0.6](https://github.com/margolisj/react-gcolor-picker/compare/0.0.5...0.0.6)

> 2 August 2023

#### [0.0.5](https://github.com/margolisj/react-gcolor-picker/compare/0.0.4...0.0.5)

> 2 August 2023

- Re-adds react-scripts + CI [`#7`](https://github.com/margolisj/react-gcolor-picker/pull/7)
- More improvements with types and package upgrades  [`#6`](https://github.com/margolisj/react-gcolor-picker/pull/6)
- Release v0.0.5 [`18d7f88`](https://github.com/margolisj/react-gcolor-picker/commit/18d7f883fbe99d32d9fc467fd7dea1e9a03c78a0)

#### [0.0.4](https://github.com/margolisj/react-gcolor-picker/compare/0.0.2...0.0.4)

> 22 May 2023

- Fix minor type issues and updates more documents. [`#5`](https://github.com/margolisj/react-gcolor-picker/pull/5)
- 0.0.3 [`#4`](https://github.com/margolisj/react-gcolor-picker/pull/4)
- Release v0.0.3 [`b2f67c5`](https://github.com/margolisj/react-gcolor-picker/commit/b2f67c545e88294e935383ca0ebc9994837ded9c)
- Release v0.0.4 [`e64dc1e`](https://github.com/margolisj/react-gcolor-picker/commit/e64dc1efc8b735864496c748d9623bffc3d01775)

#### 0.0.2

> 17 April 2023

- Clean up for NPM Publish [`#3`](https://github.com/margolisj/react-gcolor-picker/pull/3)
- Bugfix missing gradient end value [`#2`](https://github.com/margolisj/react-gcolor-picker/pull/2)
- Upgrades Packages: Storybook, react, all babel + rollup [`#1`](https://github.com/margolisj/react-gcolor-picker/pull/1)
- Fix minor bugs [`#28`](https://github.com/margolisj/react-gcolor-picker/pull/28)
- patch: remove color stop on double click [`#22`](https://github.com/margolisj/react-gcolor-picker/pull/22)
- fix: update peerDependencies [`#21`](https://github.com/margolisj/react-gcolor-picker/pull/21)
- patch: add new props [`#17`](https://github.com/margolisj/react-gcolor-picker/pull/17)
- bug: fix mousedown conflict [`#15`](https://github.com/margolisj/react-gcolor-picker/pull/15)
- fix: render components and scss styles [`#13`](https://github.com/margolisj/react-gcolor-picker/pull/13)
- patch: remove color stops [`#12`](https://github.com/margolisj/react-gcolor-picker/pull/12)
- Update prop value [`#8`](https://github.com/margolisj/react-gcolor-picker/pull/8)
- chore: fix hsl string regexp [`#5`](https://github.com/margolisj/react-gcolor-picker/pull/5)
- Default colors panel [`#4`](https://github.com/margolisj/react-gcolor-picker/pull/4)
- patch: add format component prop [`#3`](https://github.com/margolisj/react-gcolor-picker/pull/3)
- fix: react jsx runtime error [`#2`](https://github.com/margolisj/react-gcolor-picker/pull/2)
- patch: mobile touch events [`#1`](https://github.com/margolisj/react-gcolor-picker/pull/1)
- init [`8cb39a8`](https://github.com/margolisj/react-gcolor-picker/commit/8cb39a843d4c9ed310d1eb11256d9c2581fc52d1)
- Really big update for files, yarn, disables prepare currently, tool-version. [`2726075`](https://github.com/margolisj/react-gcolor-picker/commit/27260751757ce59b96a919e0a5382639a534fdb4)
- Updates storybook, adds babel presets, updates enzyme plugin for 18. [`dad43d4`](https://github.com/margolisj/react-gcolor-picker/commit/dad43d4449cffea51af68a48d65caade36399ea5)


src/utils/parseGradient.ts
import tinycolor from 'tinycolor2';

import validGradient from './validGradient';

interface IGradientStop {
  color: string;
  position?: number;
}

const LINEAR_POS = [
  { angle: '0', name: 'to top' },
  { angle: '45', name: 'to top right' },
  { angle: '45', name: 'to right top' },
  { angle: '90', name: 'to right' },
  { angle: '135', name: 'to right bottom' },
  { angle: '135', name: 'to bottom right' },
  { angle: '180', name: 'to bottom' },
  { angle: '225', name: 'to left bottom' },
  { angle: '225', name: 'to bottom left' },
  { angle: '270', name: 'to left' },
  { angle: '315', name: 'to top left' },
  { angle: '315', name: 'to left top' }
];

export default (str: string) => {
  const tinyColor = tinycolor(str);

  const defaultStops = {
    stops: [
      ['rgba(0, 0, 0, 1)', 0, 0],
      ['rgba(183, 80, 174, 0.92)', 1, 1]
    ],
    gradient: `linear-gradient(180deg, rgba(6, 6, 6, 1) 0.0%, rgba(183, 80, 174, 0.92) 100.0%)`,
    modifier: 180,
    type: 'linear'
  };

  if (str === 'transparent') {
    return defaultStops;
  }

  if (
    tinyColor.isValid() &&
    !str.trim().startsWith('radial-gradient') &&
    !str.trim().startsWith('linear-gradient')
  ) {
    const rgbaStr = tinyColor.toRgbString();

    if (rgbaStr) {
      defaultStops.stops = [
        ['rgba(0, 0, 0, 1)', 0, 0],
        [rgbaStr, 1, 1]
      ];
      defaultStops.gradient = `linear-gradient(180deg, rgba(6, 6, 6, 1) 0.0%, ${rgbaStr} 100.0%)`;
    }

    return defaultStops;
  } else {
    str = str.replace(';', '').replace('background-image:', '');
    const gradient = validGradient(str);

    let stops: Array<IGradientStop> = [];
    let angle: string = '';

    if (
      gradient === 'Failed to find gradient' ||
      gradient === 'Not correct position'
    ) {
      // console.log('Incorrect gradient value');
      return defaultStops;
    }

    if (typeof gradient !== 'string') {
      stops = gradient.stops;
      angle = gradient.angle ? gradient.angle : gradient.line;
    }

    const [, type, content] = str.match(/^(\w+)-gradient\((.*)\)$/i) || [];
    if (!type || !content) {
      // console.log('Incorrect gradient value');
      return defaultStops;
    }

    const findF = LINEAR_POS.find((item) => item.name === angle)?.angle;
    const helperAngle = type === 'linear' ? '180' : 'circle at center';
    const modifier = findF || angle || helperAngle;

    return {
      gradient: `${type}-gradient(${
        typeof gradient !== 'string' ? gradient.original : str
      })`,
      type,
      modifier:
        modifier.match(/\d+/) !== null
          ? Number(modifier.match(/\d+/)?.join(''))
          : modifier,
      stops: stops.map((stop, index: number) => {
        const formatStop = [`${stop.color}`, index];
        if (stop.position || stop.position === 0) {
          formatStop.splice(1, 0, stop.position);
        }
        return formatStop;
      })
    };
  }
};
